Renesas R-Car add CAN driver

julien-massot opened the pull request:
This PR add CAN driver for the Renesas R-Car SoC family.

It has been tested on H3ULCB and Ebisu board which are based on R-Car Gen3 SoC.
This driver use hardware FIFO mailbox for both Rx and Tx, it's possible to send and receive
standard and extended ID frames.

This driver also support loopback mode for testing purposes.

The hardware module on R-CAR Gen3 also support CAN-FD, which is not implemented by this driver.

Commit:
dts: bindings: can: Add support for Renesas R-Car

This is classic CAN bindings tested on Renesas R-Car
gen3 series.

Signed-off-by: Julien Massot <julien.massot@iot.bzh>

Commit:
drivers: can: mcp2515: move filter_match function

Move filter match function to can_utils.h, so that it
can be reused for Renesas driver.

Preserve copyright from Karsten Koenig.

Signed-off-by: Julien Massot <julien.massot@iot.bzh>

mbolivar-nordic reviewed (approved):
DT bits LGTM

alexanderwachter commented on drivers/can/can_rcar.c:
```diff
@@ -0,0 +1,1044 @@
+		msg.dlc = CAN_MAX_DLC;
+	}
+
+	for (i = 0; i < msg.dlc; i++) {
```
why no memcpy?

julien-massot replied:
I'm accessing a memory mapped device, not sure it's safe to use memcpy there.

alexanderwachter commented on drivers/can/can_rcar.c:
```diff
@@ -0,0 +1,1044 @@
+	can_rcar_rx_filter(data, &msg);
+}
+
+static void can_rcar_int_thread(const struct device *dev)
```
Is there a reason why you need a thread?
The API is defined that the callback is an ISR (Except for MCP2515).

julien-massot replied:
Idea there is to protect concurrent access of rx_callback[], and rx_callback_arg by locking data->rx_mutex.
Also for the Tx path I do lock data->inst_mutex to avoid isr, accessing and other threads accessing data->tx_head at the same time.

alexanderwachter replied:
You are writing from a thread and reading from an ISR.
The ISR cannot get interrupted.
If you have a member in the struct that can be written atomic and defines the validity, you don't need to lock.
Just write this marker at the end of an attach and at the beginning of a detach.

alexanderwachter replied:
For the ring buffer: writing the head from the ISR and tail from the thread can be done threads safe too.
You just need to add a `compiler_barier();` then moving hand/tail.

julien-massot replied:
Thanks for your suggestion, I did post a new version which get rid of this thread, and implements compiler_barrier as suggested.

alexanderwachter commented on drivers/can/can_rcar.c:
```diff
@@ -0,0 +1,1044 @@
+#ifndef CONFIG_CAN_AUTO_BUS_OFF_RECOVERY
+int can_rcar_recover(const struct device *dev, k_timeout_t timeout)
+{
+	return 0;
```
No recovery on rcar-can?

julien-massot replied:
will check

alexanderwachter commented on drivers/can/can_rcar.c:
```diff
@@ -0,0 +1,1044 @@
+	}
+
+	k_mutex_lock(&data->inst_mutex, K_FOREVER);
+	tx_cb = &data->tx_cb[data->tx_head % RCAR_CAN_FIFO_DEPTH];
```
This only works if RCAR_CAN_FIFO_DEPTH is a power of two. Otherwise you wil have a leap when the tx_head overflows

julien-massot replied:
that's true, this implementation looks a bit fragile, will try to find another trick.

julien-massot replied:
Ok did a second implementation when sending a can frame tx_head is still incremented but set to 0 in case it reaches RCAR_CAN_FIFO_DEPTH, the can_rcar_send function also increments data->tx_unsent, which is used by the tx isr handler, to know the fifo depth. That way it's easier to read and robust to any RCAR_CAN_FIFO_DEPTH value.

alexanderwachter reviewed (changes requested):
I just had a brief look.
I'll do a prorer review when I have more time :-)

Commit:
drivers: can: Add Renesas R-Car driver

This driver is for classic CAN, it makes use of CAN interface
in FIFO mode.

This driver support Standard ID as well as Extended ID.

Tested on H3ULCB, Ebisu platform, with external adapter and
in loopback mode.

Signed-off-by: Julien Massot <julien.massot@iot.bzh>

Commit:
dts: arm: rcar_gen3_cr7: add can0 node

Add can0 classic can node to the rcar_gen3_cr7 SoC series.

Signed-off-by: Julien Massot <julien.massot@iot.bzh>

Commit:
boards: arm: rcar_h3ulcb_cr7 enable can0 interface

Enable the can0 interface that can be found on H3ULCB board.

Signed-off-by: Julien Massot <julien.massot@iot.bzh>

julien-massot commented on drivers/can/Kconfig.rcar:
```diff
@@ -0,0 +1,32 @@
+
+config CAN_RCAR
+	bool "Renesas R-Car CAN Driver"
+	depends on SOC_SERIES_RCAR_GEN3
```
You're right this driver should be compatible with any soc within RCAR family.

julien-massot replied:
thanks! did replace SOC_SERIES_RCAR_GEN3 by SOC_FAMILY_RCAR.

alexanderwachter commented on dts/arm/renesas/gen3/rcar_gen3_cr7.dtsi:
```diff
@@ -85,6 +85,25 @@
+				<&cpg CPG_CORE CPG_CORE_CLK_CANFD>;
+			#address-cells = <1>;
+			#size-cells = <0>;
+			bus-speed = <500000>;
```
```suggestion
			bus-speed = <125000>;
```

alexanderwachter commented on drivers/can/can_rcar.c:
```diff
@@ -0,0 +1,1057 @@
+		return;
+	}
+
+	LOG_INF("Can state change new: %u old:%u\n", newstate, data->state);
```
```suggestion
	LOG_DBG("Can state change new: %u old:%u\n", newstate, data->state);
```

alexanderwachter commented on drivers/can/can_rcar.c:
```diff
@@ -0,0 +1,1057 @@
+	}
+#if defined(CONFIG_CAN_RX_TIMESTAMP)
+	/* read upper byte */
+	msg.timestamp = sys_read8(config->reg_addr +
+				  RCAR_CAN_MB_60 + RCAR_CAN_MB_TSH_OFFSET);
+	msg.timestamp = msg.timestamp << 8;
+	/* and then read lower byte */
+	msg.timestamp |= sys_read8(config->reg_addr +
```
You are writing twice here

alexanderwachter replied:
Ah, now I saw what you tried to to. Sorry, my bad. that was OK.

alexanderwachter commented on drivers/can/can_rcar.c:
```diff
@@ -0,0 +1,1056 @@
+			return 0;
+		}
+	}
+	return -ETIME;
```
s/-ETIME/CAN_TIMEOUT/

It seems you are returning this directly.
According to the API it should be CAN_TIMEOUT.
Sorry, did not catch this in the last review

julien-massot replied:
Thanks did change these return

alexanderwachter commented:
I assume all test are passing?

julien-massot commented:
> I assume all test are passing?

All Can tests do pass excepted 
zephyr/tests/subsys/canbus/isotp/implementation
This is due to the fact that R-Car CAN controller send Rx isr before sending the Tx isr,
which totally confuse the Zephyr ISOTP state machine. ISOTP test suite can pass if one add manually a
delay in R-Car CAN ISR handler for testing purpose, so I guess the implementation of this driver is correct.

alexanderwachter reviewed (approved):
Looks good now!
